spring-context/src/main/java/org/springframework/scheduling/annotation/ScheduledAnnotationReactiveSupport.java

```diff
@@ -0,0 +1,209 @@
+ * Copyright 2002-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.scheduling.annotation;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.time.Duration;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.reactivestreams.Publisher;
+import reactor.core.Disposable;
+import reactor.core.Disposables;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import org.springframework.aop.support.AopUtils;
+import org.springframework.core.CoroutinesUtils;
+import org.springframework.core.KotlinDetector;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.ReflectionUtils;
+
+/**
+ * Helper class for @{@link ScheduledAnnotationBeanPostProcessor} to support reactive cases
+ * without a dependency on optional classes.
+ * @author Simon Baslé
+ * @since 6.1.0 //FIXME
+ */
+abstract class ScheduledAnnotationReactiveSupport {
+
+	static final boolean publisherPresent = ClassUtils.isPresent(
+			"org.reactivestreams.Publisher", ScheduledAnnotationReactiveSupport.class.getClassLoader());
+
+	static final boolean reactorPresent = ClassUtils.isPresent(
+			"reactor.core.publisher.Flux", ScheduledAnnotationReactiveSupport.class.getClassLoader());
+
+	static final boolean coroutinesReactorPresent = ClassUtils.isPresent(
+			"kotlinx.coroutines.reactor.MonoKt", ScheduledAnnotationReactiveSupport.class.getClassLoader());
+
+	/**
+	 * Checks that if the method is reactive, it can be scheduled. Methods are considered
+	 * eligible for reactive scheduling if they either return an instance of
+	 * {@code Publisher} or are a Kotlin Suspending Function. If the method isn't matching
+	 * these criteria then this check returns {@code false}.
+	 * <p>For reactive scheduling, Reactor MUST be present at runtime. In the case of
+	 * Kotlin, the Coroutine {@code kotlinx.coroutines.reactor} bridge MUST also be
+	 * present at runtime (in order to invoke suspending functions as a {@code Mono}).
+	 * Provided that is the case, this method returns {@code true}. Otherwise, it throws
+	 * an {@code IllegalStateException}.
+	 * @throws IllegalStateException if the method is reactive but Reactor and/or the
+	 * Kotlin coroutines bridge are not present at runtime
+	 */
+	static boolean isReactive(Method method) {
```

Comment:
Couldn't we use `ReactiveAdapterRegistry#getSharedInstance` to support various other non `Publisher` reactive types?

Reply:
I was under the impression that the canonical `ReactiveAdapterRegistry` isn't necessarily the `getSharedInstance` one but a bean one. When I started on this PR I initially tried to use a `ReactiveAdapterRegistry` from the context, but at the time the processing ran it wasn't available (which makes it difficult to detect arbitrary "reactive-convertible" types).

Reply:
Given the fact that this feature needs to be implemented at `BeanPostProcessor` level and the fact that `ReactiveAdapterRegistry#getSharedInstance()` is pretty useful and flexible out of the box via its classpath detection mechanism, I think for this use case it is reasonable to have the limitation that it will not be customizable from the application context. Maybe worth documenting both at Javadoc and reference documentation level.